File: leetcode_75/binary_tree_DFS/leaf-similar_trees_872.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def getLeaves(node):
            if not node:
                return []
            if node.left is None and node.right is None:
                return [node.val]
            return getLeaves(node.left) + getLeaves(node.right)

        return getLeaves(root1) == getLeaves(root2)
        # r1_temp = []
        # if root1.left is None and root1.right is None:
        #     r1_temp.append(root1.val)
        # if root1.left:
        #     self.leafSimilar(root1.left)
        # if root1.right:
        #     self.leafSimilar(root1.right)

File: leetcode_75/binary_tree_DFS/leaf-similar_trees_872.java
class SolutionWithStream {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        var leaves1 = getLeaves(root1).toList();
        var leaves2 = getLeaves(root2).toList();
        return leaves1.equals(leaves2);
    } // equals 這是比較內容物，不是比較資料地址

    private Stream<Integer> getLeaves(TreeNode node) {
        if (node == null)
            return Stream.empty();
        if (node.left == null && node.right == null)
            return Stream.of(node.val);
        return Stream.concat(getLeaves(node.left), getLeaves(node.right));
    } // concat 是兩個list的合併
}

class SolutionWithOri {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> leaves1 = new ArrayList<>();
        List<Integer> leaves2 = new ArrayList<>();

        getLeaves(root1, leaves1);
        getLeaves(root2, leaves2);

        return leaves1.equals(leaves2);
    }

    private void getLeaves(TreeNode node, List<Integer> leaves) {
        if (node == null)
            return;

        if (node.left == null && node.right == null) {
            leaves.add(node.val);
            return;
        }

        getLeaves(node.left, leaves);
        getLeaves(node.right, leaves);
    }
}

// class Solution {
// public boolean leafSimilar(TreeNode root1, TreeNode root2) {
// public int[] getLeave(TreeNode node) {
// if (node == null) return int[];
// if (node.left == null & node.right == null) return [node.val];
// return getLeaves(node.left) + getLeaves(node.right);
// }
// return getLeave(root1) == getLeave(root2);
// }
// }